cmd/terminal.go

```diff
@@ -0,0 +1,367 @@
+// Copyright 2021 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+package cmd
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"github.com/vmihailenco/msgpack"
+	"golang.org/x/crypto/ssh/terminal"
+	"golang.org/x/sys/unix"
+	"golang.org/x/term"
+
+	"github.com/mendersoftware/go-lib-micro/ws"
+	wsshell "github.com/mendersoftware/go-lib-micro/ws/shell"
+
+	"github.com/mendersoftware/mender-cli/log"
+)
+
+const (
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Time allowed to read the next pong message from the peer.
+	pongWait = 1 * time.Minute
+
+	// protocols
+	httpsProtocol = "https"
+	httpProtocol  = "http"
+	wssProtocol   = "wss"
+	wsProtocol    = "ws"
+)
+
+var terminalCmd = &cobra.Command{
+	Use:   "terminal DEVICE_ID",
+	Short: "Access a device's remote terminal",
+	Args:  cobra.ExactArgs(1),
+	Run: func(c *cobra.Command, args []string) {
+		cmd, err := NewTerminalCmd(c, args)
+		CheckErr(err)
+		CheckErr(cmd.Run())
+	},
+}
+
+func getWebSocketScheme(scheme string) string {
+	if scheme == httpsProtocol {
+		scheme = wssProtocol
+	} else if scheme == httpProtocol {
+		scheme = wsProtocol
+	}
+	return scheme
+}
+
+// TerminalCmd handles the terminal command
+type TerminalCmd struct {
+	server     string
+	skipVerify bool
+	deviceID   string
+	sessionID  string
+	running    bool
+	stop       chan bool
+	err        error
+}
+
+// NewTerminalCmd returns a new TerminalCmd
+func NewTerminalCmd(cmd *cobra.Command, args []string) (*TerminalCmd, error) {
+	server, err := cmd.Flags().GetString(argRootServer)
+	if err != nil {
+		return nil, err
+	}
+
+	skipVerify, err := cmd.Flags().GetBool(argRootSkipVerify)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TerminalCmd{
+		server:     server,
+		skipVerify: skipVerify,
+		deviceID:   args[0],
+		stop:       make(chan bool),
+	}, nil
+}
+
+// Run executes the command
+func (c *TerminalCmd) Run() error {
```

Comment:
I'm surprised, doesn't this function trigger the cyclomatic complexity check? It is very long, and I think it makes sense to break it up.

Reply:
Do we really need to? I know it is long, but it is extremely linear. Breaking it down requires additional struct-level variables, and makes it less readable, IMHO.

```
mender-cli (terminal) $ gocyclo cmd/
17 cmd (*TerminalCmd).Run cmd/terminal.go:100:1
```

Reply:
I would at least consider putting the main loop in a sub function and run it there, since this is doing a very different job than the rest of the function.

Reply:
Done: https://github.com/mendersoftware/mender-cli/pull/97/commits/7a46802f03430c7168e551a05f9abad23fa24cbf

Reply:
Thanks, that makes it easier to read!